File: Python3.x/34-Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        start, end, length = 0, len(nums) - 1, len(nums) - 1
        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target: break
            if nums[mid] < target: start = mid + 1
            else: end = mid - 1
        else: return [-1, -1]
        start = end = mid
        while start >= 0 and nums[start] == target: start -= 1
        while end <= length and nums[end] == target: end += 1
        return [start + 1, end - 1]


'''
先二分搜索找到目标值
然后向左和向右搜索
'''

File: Algorithms/C++/34-Find_First_and_Last_Position_of_Element_in_Sorted_Array.cpp
# Runtime: 8 ms, faster than 86.80% of C++ online submissions for Find First and Last Position of Element in Sorted Array.
# Memory Usage: 10.1 MB, less than 98.57% of C++ online submissions for Find First and Last Position of Element in Sorted Array.
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> res = {-1, -1};
        if (nums.size() == 0)
            return res;
        int l = 0, r = nums.size() - 1, idx = NULL;
        while (l <= r) { // l and r is region pointer in here
            int mid = (l + r) / 2;
            if (nums[mid] == target) {
                idx = mid;
                break;
            } else if (nums[mid] > target)
                r = mid - 1;
            else
                l = mid + 1;
        }
        if (nums[idx] == target) { // judge find target index
            l = idx;
            r = idx;
            while (l > 0)
                if (target == nums[l - 1])
                    --l;
                else
                    break;
            while (r < nums.size() - 1)
                if (target == nums[r + 1])
                    ++r;
                else
                    break;
            res[0] = l;
            res[1] = r;
        }
        return res;
    }
};
